lib/charms/data_platform_libs/v0/data_interfaces.py

```diff
@@ -534,14 +755,65 @@ def _diff(self, event: RelationChangedEvent) -> Diff:
         """
         return diff(event, self.local_unit)
 
-    @staticmethod
-    def _is_resource_created_for_relation(relation: Relation) -> bool:
+    def get_relation_secret(
+        self, relation_id: int, label: str, relation_name: Optional[str] = None
+    ) -> Optional[Secret]:
+        """Retrieve a Juju Secret that's been stored in the relation databag."""
+        if not relation_name:
+            relation_name = self.relation_name
+
+        if not self.secrets.get(relation_id, {}).get(label) and self.secrets_enabled:
+            relation = self.charm.model.get_relation(relation_name, relation_id)
+            if not relation or not relation.app:
+                return
+            if secret_id := relation.data[relation.app].get(f"secret-{label}"):
+                self.secrets.setdefault(relation_id, {})[label] = SecretCache(
+                    self.charm, secret_id
+                ).get_secret()
+        return self.secrets.get(relation_id, {}).get(label)
+
+    def get_relation_fields(
+        self, relation_id: int, fields: List[str], relation_name: Optional[str] = None
+    ) -> Dict[str, str]:
+        """Get the value of a field not caring whether it's a secret or not."""
+        if not relation_name:
+            relation_name = self.relation_name
+
+        result = {}
+        relation = self.get_relation(relation_name, relation_id)
+        relation_secret_fields = relation.data.get(
+            self.local_app, {}  # pyright: ignore [reportGeneralTypeIssues]
+        ).get("secret_fields")
+
+        normal_fields = fields
+        if relation_secret_fields:
+            relation_secret_fields = relation_secret_fields.split(" ")
+            normal_fields = set(fields) - set(relation_secret_fields)
+            secret_fields = set(fields) - set(normal_fields)
+
+            if secret_fields and self.secrets_enabled:
+                label_sorted_content = self._create_label_sorted_content(fields)
+                for label in label_sorted_content:
+                    if secret := self.get_relation_secret(relation_id, label):
+                        if secret_data := secret.get_content():
+                            result.update(
+                                {k: v for k, v in secret_data.items() if k in secret_fields}
+                            )
```

Comment:
stylistic nit: i feel like the cyclomatic complexity of this code block is too high -> suggest trying to move block 794-801 outside of the if statement on line 789

```suggestion
        normal_fields, secret_fields = fields, []
        if relation_secret_fields:
            relation_secret_fields = relation_secret_fields.split(" ")
            normal_fields = set(fields) - set(relation_secret_fields)
            secret_fields = set(fields) - set(normal_fields)

        if secret_fields and self.secrets_enabled:
            label_sorted_content = self._create_label_sorted_content(fields)
            for label in label_sorted_content:
                if (secret := self.get_relation_secret(relation_id, label)) and (secret_data := secret.get_content()(:
                    result.update(
                        {k: v for k, v in secret_data.items() if k in secret_fields}
                    )
```